subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy

```diff
@@ -187,67 +183,247 @@ class JavaToolchainBuildOperationsIntegrationTest extends AbstractIntegrationSpe
+                    languageVersion = JavaLanguageVersion.of(${jdkMetadata.languageVersion.majorVersion})
+                }
+            }
+
+            def javaExecutable = javaToolchains.launcherFor {
+                languageVersion = JavaLanguageVersion.of(${jdkMetadata.languageVersion.majorVersion})
+            }.get().executablePath
+
+            test {
+                executable = javaExecutable
+            }
+        """
+
+        file("src/test/java/FooTest.java") << """
+            public class FooTest {
+                @org.junit.Test
+                public void test() {}
+            }
+        """
+
+        def task = ":test"
+
+        when:
+        runWithInstallation(jdkMetadata, task)
+//        def events = eventsFor(task)
+        then:
+        executedAndNotSkipped(task)
+        // TODO: uncomment this when fixed
+//        assertToolchainUsages(events, jdkMetadata, "JavaLauncher")
+
+        when:
+        runWithInstallation(jdkMetadata, task)
+//        events = eventsFor(task)
+        then:
+        skipped(task)
+        // TODO: uncomment this when fixed
+//        assertToolchainUsages(events, jdkMetadata, "JavaLauncher")
+    }
+
+    def "emits toolchain usages when configuring toolchains for Kotlin plugin for '#task' task"() {
+        JvmInstallationMetadata jdkMetadata = AvailableJavaHomes.getJvmInstallationMetadata(AvailableJavaHomes.differentVersion)
+
+        // override setup
+        buildFile.text = """
+            buildscript {
+                ${mavenCentralRepository()}
+                dependencies { classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:1.7.10" }
+            }
+
+            apply plugin: "kotlin"
+
+            ${mavenCentralRepository()}
+            dependencies {
+                testImplementation 'junit:junit:4.13'
+            }
+
+            kotlin {
+                jvmToolchain {
+                    languageVersion = JavaLanguageVersion.of(${jdkMetadata.languageVersion.majorVersion})
+                }
+            }
+        """
+
+        file("src/main/kotlin/Foo.kt") << """
+            class Foo {
+                fun random() = 4
+            }
+        """
+
+        file("src/test/kotlin/FooTest.kt") << """
+            class FooTest {
+                @org.junit.Test
+                fun test() {}
+            }
+        """
+
+        when:
+        runWithInstallation(jdkMetadata, task)
+        def events = eventsFor(task)
+        then:
+        executedAndNotSkipped(task)
+        // The tool is a launcher, because kotlin runs own compilation in a Java VM
+        assertToolchainUsages(events, jdkMetadata, "JavaLauncher")
+
+        when:
+        runWithInstallation(jdkMetadata, task)
+        events = eventsFor(task)
+        then:
+        skipped(task)
+        assertToolchainUsages(events, jdkMetadata, "JavaLauncher")
+
+        where:
+        task             | _
+        ":compileKotlin" | _
+        ":test"          | _
```

Comment:
How is test related here?

Reply:
It checks that even though we only configure the toolchain within the `kotlin` block, it actually affects the java launcher selected by the `test` task.